pkg/plugins/runtime/k8s/controllers/gatewayapi/policy/policy_test.go

```diff
@@ -0,0 +1,176 @@
+package policy_test
+
+import (
+	"context"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	kube_meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kube_runtime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	kube_client_fake "sigs.k8s.io/controller-runtime/pkg/client/fake"
+	gatewayapi "sigs.k8s.io/gateway-api/apis/v1alpha2"
+
+	"github.com/kumahq/kuma/pkg/plugins/bootstrap/k8s"
+	mesh_k8s "github.com/kumahq/kuma/pkg/plugins/resources/k8s/native/api/v1alpha1"
+	"github.com/kumahq/kuma/pkg/plugins/runtime/k8s/controllers/gatewayapi/policy"
+)
+
+var k8sScheme *kube_runtime.Scheme
+
+var _ = BeforeSuite(func() {
+	var err error
+	k8sScheme, err = k8s.NewScheme()
+	Expect(err).NotTo(HaveOccurred())
+})
+
+const (
+	defaultNs = "default"
+	otherNs   = "other"
+)
+
+var (
+	httpRouteGK  = schema.GroupKind{Group: gatewayapi.GroupName, Kind: "HTTPRoute"}
+	simplePolicy = gatewayapi.ReferencePolicy{
+		ObjectMeta: kube_meta.ObjectMeta{
+			Name:      "basic",
+			Namespace: otherNs,
+		},
+		Spec: gatewayapi.ReferencePolicySpec{
+			From: []gatewayapi.ReferencePolicyFrom{
+				{
+					Group:     gatewayapi.Group(gatewayapi.GroupName),
+					Kind:      gatewayapi.Kind("HTTPRoute"),
+					Namespace: gatewayapi.Namespace(defaultNs),
+				},
+			},
+			To: []gatewayapi.ReferencePolicyTo{
+				{
+					Group: gatewayapi.Group(""),
+					Kind:  gatewayapi.Kind("Service"),
+				},
+			},
+		},
+	}
+
+	// References
+	coreGroup = gatewayapi.Group("")
+	svcKind   = gatewayapi.Kind("Service")
+	somePort  = gatewayapi.PortNumber(80)
+
+	toDefaultNs  = gatewayapi.Namespace(defaultNs)
+	toDefaultSvc = gatewayapi.BackendObjectReference{
+		Group:     &coreGroup,
+		Kind:      &svcKind,
+		Name:      "svc",
+		Namespace: &toDefaultNs,
+		Port:      &somePort,
+	}
+
+	toOtherNs  = gatewayapi.Namespace(otherNs)
+	toOtherSvc = gatewayapi.BackendObjectReference{
+		Group:     &coreGroup,
+		Kind:      &svcKind,
+		Name:      "svc",
+		Namespace: &toOtherNs,
+		Port:      &somePort,
+	}
+
+	kumaGroup          = gatewayapi.Group(mesh_k8s.GroupVersion.Group)
+	externalSvcKind    = gatewayapi.Kind("ExternalService")
+	toOtherExternalSvc = gatewayapi.BackendObjectReference{
+		Group:     &kumaGroup,
+		Kind:      &externalSvcKind,
+		Name:      "ext-svc",
+		Namespace: &toOtherNs,
+	}
+)
```

Comment:
Going to leave it because IMO increasing the nesting makes things harder to read for little benefit, the test cases are the only things in the `policy_test` package anyway.